account_multicurrency_revaluation/model/account.py

```diff
@@ -0,0 +1,121 @@
+            self.env.ref("account.data_account_type_payable").id,
+            self.env.ref("account.data_account_type_liquidity").id,
+        ]
+
+    @api.multi
+    @api.onchange('user_type_id')
+    def _onchange_user_type_id(self):
+        for rec in self:
+            if rec.user_type_id.id in rec._get_revaluation_account_types():
+                rec.currency_revaluation = True
+
+    @api.multi
+    def _revaluation_query(self, revaluation_date):
+
+        tables, where_clause, where_clause_params = \
+            self.env['account.move.line']._query_get()
+
+        query = ("with amount as ( SELECT aml.account_id, aml.partner_id, "
+                 "aml.currency_id, aml.debit, aml.credit, aml.amount_currency "
+                 "FROM account_move_line aml LEFT JOIN "
+                 "account_partial_reconcile aprc ON (aml.balance < 0 "
+                 "AND aml.id = aprc.credit_move_id) LEFT JOIN "
+                 "account_move_line amlcf ON (aml.balance < 0 "
+                 "AND aprc.debit_move_id = amlcf.id "
+                 "AND amlcf.date < %s ) LEFT JOIN "
+                 "account_partial_reconcile aprd ON (aml.balance > 0 "
+                 "AND aml.id = aprd.debit_move_id) LEFT JOIN "
+                 "account_move_line amldf ON (aml.balance > 0 "
+                 "AND aprd.credit_move_id = amldf.id "
+                 "AND amldf.date < %s ) "
+                 "WHERE aml.account_id IN %s "
+                 "AND aml.date <= %s "
+                 "AND aml.currency_id IS NOT NULL "
+                 "GROUP BY aml.id "
+                 "HAVING aml.full_reconcile_id IS NULL "
+                 "OR (MAX(amldf.id) IS NULL AND MAX(amlcf.id) IS NULL)"
+                 ") SELECT account_id as id, partner_id, currency_id, " +
+                 ', '.join(self._sql_mapping.values()) +
+                 " FROM amount " +
+                 (("WHERE " + where_clause) if where_clause else " ") +
+                 " GROUP BY account_id, currency_id, partner_id")
+
+        params = []
+        params.append(revaluation_date)
+        params.append(revaluation_date)
+        params.append(tuple(self.ids))
+        params.append(revaluation_date)
+        params += where_clause_params
+        return query, params
+
+    @api.multi
+    def compute_revaluations(self, revaluation_date):
+        accounts = {}
+        # compute for each account the balance/debit/credit from the move lines
+        query, params = self._revaluation_query(revaluation_date)
+        self.env.cr.execute(query, params)
+
+        lines = self.env.cr.dictfetchall()
+
+        for line in lines:
+            # generate a tree
+            # - account_id
+            # -- currency_id
+            # --- partner_id
+            # ----- balances
+            account_id, currency_id, partner_id = \
+                line['id'], line['currency_id'], line['partner_id']
+
+            accounts.setdefault(account_id, {})
```

Comment:
you can use `collections.defaultdict` :wink:

Reply:
I'm pretty sure nesting defaultdicts will make it less readable :\